### ISSUE TITLE
[question] Clang command line arguments and custom libc++

### ISSUE THREAD

**Issue by skipfowler** [conan-io/conan#13591]
### What is your question?

TL;DR: Can Conan provide argument to invocations of clang? Specifically arguments for using a custom built libc++ in a non-standard location.
___
Conan Community,

I am working in an environment where `/usr` is locked down by IT. C++ standard library changes were pushed by IT breaking building products on Linux. I manage our on-site Conan repositories and build processes. The team has decided to mitigate risk from IT by standardizing our compiler installation to clang for Linux and Windows. Linux has taken priority given our broken product.

I was able to build clang on Linux with correct enough [flavors](https://blog.conan.io/2022/10/13/Different-flavors-Clang-compiler-Windows.html). I built runtimes and linked against them explicitly [using a custom built libc++](https://libcxx.llvm.org/UsingLibcxx.html#using-a-custom-built-libc). Using `ldd` I was able to verify that the program was linking against the custom libc++ and libc++abi.

I attempted to extend that idea to our on-site Conan repositories. I was unsuccessful at finding a way to provide arguments to clang via Conan. I did find [global.conf settings like tools.build:cxxflags](https://docs.conan.io/1/reference/config_files/global_conf.html#tools-configurations), but the arguments didn't seem to propagate to clang invocations from build systems. Ideally, I would like to provide the command line arguments to clang to use the custom libc++ and libc++abi, but I haven't figured it out yet.

I was able to resolve a number of issues with environment variables. I will enumerate the specifics.

1. Set up Conan profile to use settings for custom compiler
2. Setting `CC` and `CXX` to the desired compiler for ensuring CMake finds the custom compiler
3. Setting `CPATH` to the location of headers for the custom libc++ as a substitute for `-isystem <install>/include/c++/v1`
4. Setting `LIBRARY_PATH` to the location of the custom libc++ library as a substitute for `-L <install>/lib`
5. Add environment variables to the Conan profile
6. Alias clang and clang++ to the custom compiler helping build systems find the correct compiler

The steps above allowed many dependencies to build using Conan. The dependencies that did not build produced errors I could verify with my test program. I had a duplicate definition issue when I removed the [-nostdinc++](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang1-nostdinc) argument. I had unresolved external symbol issues when I removed the [-Wl,-rpath,<install>/lib](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-wl-arg-arg2) argument.

I would like to know if [Conan's tool configurations](https://docs.conan.io/1/reference/config_files/global_conf.html#tools-configurations) can provide arguments to the compiler. I would like to know about strategies to manage the compiler using Conan. I would like to know anything I am overlooking. The solution may be twisting IT's arm to get updated compilers in typical locations.

Thank you!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @skipfowler 

Thanks for your question and your detailed explanation.

> The steps above allowed many dependencies to build using Conan.

From this I understand that you might be using some recipes from ConanCenter, and some work and some don't. 
This is kind of expected, unfortunately, many open source libraries build-systems have complicated, arcane build systems, often with many limitations, hardcoded assumptions about toolchains, etc. This makes almost impossible to address this from Conan tool, and it might require some extensive work in those recipes, patching and fixing them.

I think we should look at this first from the Conan client. For that purpose, reporting against the predefined templates, preferibly in Conan 2.0: ``conan new cmake_lib -d name=pkg -d version=0.1`` (and msbuild, autotools, meson, etc) would be the way to go, to guarantee the Conan client is ready and works, before approaching the different recipes in ConanCenter.

**Comment by skipfowler** [comment]
Hi @memsharded,

I am using some recipes from ConanCenter. There is value in checking `conan new` against varied templates. I appreciate the feedback and will report findings.

Thank you,
Skip

**Comment by memsharded** [comment]
Closing the ticket as staled, never followed up, but please create new tickets for any further question or issue, thanks for the feedback!

Also, Conan 2 has provided new mechanisms that can implement this:

- Disabling automatic addition of flags by some toolchains like AutotoolsToolchain via ``tools.gnu:disable_flags``
- Possibility to inject arbitrary flags, via confs like ``tools.build:cxxflags``.
